### ISSUE TITLE
[all] sym-links may not be created in non-clean builds 

### ISSUE THREAD

**Issue by jamuir** [sonyxperiadev/bug_tracker#253]
**Platform:**
**Device:** 
**Kernel version:** 
**Android version:** 8.1

**Description**
Non-clean or incremental builds, are builds where the aosp build system figures out what has changed and then only rebuilds what is required (these are builds where you do not run "make clean" immediately beforehand).

The makefiles in `device-sony-common` contain several isolated `$(shell ...)` commands used to create sym-links.  In non-clean builds, sometimes those commands are not executed and the sym-links do not get created.  

https://github.com/sonyxperiadev/device-sony-common/blob/o-mr1/hardware/tftp/Android.mk

**Symptoms**
The resulting image does not boot.

**How to reproduce**
I don't yet have a precise way to reproduce this.  Sometimes I switch lunch configurations between g8141 and g8142 -- maybe that is somehow causing the shell-commands not to be executed.

Roughly:  build your tree, change some code, run "make installclean", build again, check for missing sym-links:
```bash
$ find $OUT/system -type l
```

**Comment by jamuir** [comment]
The isolated `$(shell ...)` commands are in Android.mk files, which get converted to soong during the build.  It might be that the conversion tool is dropping those commands.  Or maybe the build system is executing the commands at the wrong time.

If the commands to create the sym-links were re-written using normal target-prerequisite style, then I suspect this problem would go away.

Note that there is already a helper-function inside "build" to create sym-links:

https://android.googlesource.com/platform/build/+/android-8.1.0_r46/core/definitions.mk#2883

**Comment by jerpelea** [comment]
@jamuir any updates on this one ?

**Comment by jamuir** [comment]
Yes.  I am able to reproduce this by switching between maple variants:

```bash
$ lunch aosp_g8141-userdebug
$ make clean
$ make -j8 &> make.log
$ find $OUT/system -type l &> list-of-sym-links-8141.txt

$ lunch aosp_g8142-userdebug
$ make installclean
$ make -j8 &> make.log
$ find $OUT/system -type l &> list-of-sym-links-8142.txt

$ ls -l list-of-sym-links-814*.txt
-rw-rw-r-- 1 jamuir jamuir 29107 Dec 10 13:30 list-of-sym-links-8141.txt
-rw-rw-r-- 1 jamuir jamuir 23944 Dec 10 13:40 list-of-sym-links-8142.txt

$ wc --lines list-of-sym-links-814*.txt
  401 list-of-sym-links-8141.txt
  340 list-of-sym-links-8142.txt
  741 total
```
You can see that there are 61 sym-links missing from the 8142 image.

[list-of-sym-links-8141.txt](https://github.com/sonyxperiadev/bug_tracker/files/2664487/list-of-sym-links-8141.txt)
[list-of-sym-links-8142.txt](https://github.com/sonyxperiadev/bug_tracker/files/2664490/list-of-sym-links-8142.txt)

**Comment by MartinX3** [comment]
Facing the same issue.
Now I understand, why my Single Sim build boots into desktop before several services crash and it shuts down itself, while my dual sim build, which I compile after that, just stay at the SONY logo and can't init the GPU due to missing symlinks in the vendor partition linking to the gpu drivers in the OEM partition.

**Comment by MarijnS95** [comment]
These `shell` commands are only executed when soong/ninja discovers build files. There is no make target that the build system can execute at any logical time during the build.

The reason it always works for @MartinX3 on Omnirom is because they regenerate the intermediate build script when the time changes (so, every rebuild). It'll re-evaluate the file and create symlinks again. This is probably a _"better safe than sorry"_ hack.

The solution is to create a target with `LOCAL_POST_INSTALL_CMD`, where all the linker commands reside. The function @jamuir linked isn't going to work because it adds an explicit dependency on the source - which are not available during build as these reside in the (separately flashed) odm image.

If you want, I can set this up later this evening. Let me know so we aren't doing duplicate work.

**Comment by jamuir** [comment]
thanks for your updates.

@MarijnS95 :  I haven't done any work on a fix -- go for it.

I'm interested to see `LOCAL_POST_INSTALL_CMD` in action :-)

**Comment by MarijnS95** [comment]
@0xlne I'm sure Alin does clean builds (`rm -r out`) which regenerate soong/ninja intermediate files, in turn creating the symlinks. While it could be a bash/locale issue, that doesn't explain why omni always works (which uses the exact same makefile for symlink creation). All I can think of now is that @MartinX3 did "partial" rebuilds, for example removing just `out/target/product/akari` instead of everything. If he'd remove the out directory after running lunch/brunch, those build files will be regenerated anyway when running make.

In any case, the symlink creation problem (topic of this issue) needed to be fixed anyway, and I hope it also fixes Martins issue at #283. @MartinX3 and @jamuir: tagging you both to check https://github.com/sonyxperiadev/device-sony-common/pull/552.

**Comment by MartinX3** [comment]
Yes, I build a device and then just lunch another device to build it.
The initial build takes 3 hours on my laptop. :)
